simpl/db/mongodb.py

```diff
@@ -0,0 +1,554 @@
+
+        Note: close the cursor after using it if you don't exhaust it
+        """
+        projection = {'_id': False}
+        if fields:
+            projection.update({field: True for field in fields})
+        results = self._collection.find(spec=kwargs, fields=projection)
+        if sort:
+            sort_pairs = sort[:]
+            for index, field in enumerate(sort):
+                if field[0] == "-":
+                    sort_pairs[index] = (field[1:], pymongo.DESCENDING)
+                else:
+                    sort_pairs[index] = (field, pymongo.ASCENDING)
+            results.sort(sort_pairs)
+        results.skip(offset or 0).limit(limit or 0)
+        return results
+
+    def delete(self, key):
+        """Delete a document by id."""
+        assert key, "A key must be supplied for delete operations"
+        self._collection.remove(spec_or_id={'_id': key})
+        LOG.debug("DB REMOVE: %s.%s", self.collection_name, key)
+
+    def exists(self, key):
+        """True if a document exists."""
+        try:
+            return self._collection.find_one({'_id': key}) is not None
+        except StopIteration:
+            return False
+
+    def get(self, key):
+        """Get a document by id."""
+        doc = self._collection.find_one({'_id': key})
+        if doc:
+            doc.pop('_id')
+            return doc
+
+
+class KeyTransform(SONManipulator):
+
+    """Transforms keys going to database and restores them coming out.
+
+    This allows keys with dots in them to be used (but does break searching on
+    them unless the find command also uses the transform).
+
+    Example & test:
+        # To allow `.` (dots) in keys
+        import pymongo
+        client = pymongo.MongoClient("mongodb://localhost")
+        db = client['delete_me']
+        db.add_son_manipulator(KeyTransform(".", "_dot_"))
+        db['mycol'].remove()
+        db['mycol'].update({'_id': 1}, {'127.0.0.1': 'localhost'}, upsert=True,
+                           manipulate=True)
+        print db['mycol'].list().next()
+        print db['mycol'].list({'127_dot_0_dot_0_dot_1': 'localhost'}).next()
+
+    Note: transformation could be easily extended to be more complex.
+    """
+
+    def __init__(self, replace, replacement):
+        """Initialize KeyTransform."""
+        self.replace = replace
+        self.replacement = replacement
+
+    def transform_key(self, key):
+        """Transform key for saving to database."""
+        return key.replace(self.replace, self.replacement)
+
+    def revert_key(self, key):
+        """Restore transformed key returning from database."""
+        return key.replace(self.replacement, self.replace)
+
+    def transform_incoming(self, son, collection):
+        """Recursively replace all keys that need transforming."""
+        return self._transform_incoming(copy.deepcopy(son), collection)
+
+    def _transform_incoming(self, son, collection, skip=0):
```

Comment:
Confession: I contributed to this function, and I'm wary of it. On rare occasions I still see the following error come through on airbrake for our applications using this code:

```
 InvalidDocument: key 'akey.with.dots' must not contain '.'
```